### PR TITLE
Dynamic schema functions type check

### PR DESCRIPTION
**Pull request by jsingle**
@tleach

### CODE REVIEW

File: tests/schemer/sample.py [hunk 1]
```diff
@@ -36,6 +36,16 @@ def get_author_schema(document):
...
         return name_schema
 
+website_schema = Schema({
+    "url": {"type": basestring, "required": True},
+    "name": {"type": basestring, "required": True}
+    })
+
+def get_website_schema(document):
+    if isinstance(document, list):
+        return list
```
Comment (tleach) [inline tests/schemer/sample.py]: Maybe return an `Array` here which validates each member adheres to the `website_schema`??

File: tests/schemer/schema_test.py [hunk 2]
```diff
@@ -210,9 +210,11 @@ def test_set_non_nullable_field_to_none(self):
         self.document_1['external_code'] = None
         self.assert_document_paths_invalid(self.document_1, ['external_code'])
 
-    def test_incorrect_type(self):
+    def test_dynamic_function_exception(self):
         self.document_1['author'] = 33
-        self.assert_document_paths_invalid(self.document_1, ['author'])
+        with self.assertRaises(SchemaFormatException) as cm:
+            blog_post_schema.validate(self.document_1)
+        self.assertEqual('author', cm.exception.path)
```
Comment (tleach) [inline tests/schemer/schema_test.py]: Can we also test what happens when the dynamic function succeeds in returning a type, but the actual value being validated then fails the type check.

File: schemer/__init__.py [hunk 3]
```diff
@@ -225,13 +225,12 @@ def _validate_value(self, value, field_spec, path, errors):
...
+            try:
                 field_type = field_type(value)
-                if not isinstance(field_type, type) and not isinstance(field_type, Schema):
-                    raise SchemaFormatException("Dynamic schema function did not return a type at path {}", path)
-            else:
-                errors[path] = "Field type of function expects a value of type dict"
-                return
+            except Exception as e:
+                raise SchemaFormatException("Dynamic schema function raised exception: {}".format(str(e)), path)
+            if not (isinstance(field_type, type) or isinstance(field_type, Schema)):
```
Comment (tleach) [inline schemer/__init__.py]: This should probably support `Array`s as well.

File: tests/schemer/schema_test.py (hunk 2 above)
Comment (jsingle) [inline tests/schemer/schema_test.py]: added

File: schemer/__init__.py [hunk 4]
```diff
@@ -225,13 +225,12 @@ def _validate_value(self, value, field_spec, path, errors):
...
+            try:
                 field_type = field_type(value)
-                if not isinstance(field_type, type) and not isinstance(field_type, Schema):
-                    raise SchemaFormatException("Dynamic schema function did not return a type at path {}", path)
-            else:
-                errors[path] = "Field type of function expects a value of type dict"
-                return
+            except Exception as e:
+                raise SchemaFormatException("Dynamic schema function raised exception: {}".format(str(e)), path)
+            if not (isinstance(field_type, type) or isinstance(field_type, Schema) or isinstance(field_type, Array)):
```
Comment (tleach) [inline schemer/__init__.py]: If think you can pass a tuple of classes to `isinstance`

**Pull request closed without merging**